gix-fsck/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+//! A library for performing object database integrity and connectivity checks
+#![deny(rust_2018_idioms)]
+
+use gix_hash::ObjectId;
+use gix_hashtable::HashSet;
+use gix_object::{Kind, tree::EntryMode};
+use gix_odb::{Handle, FindExt, pack::Find};
+
+pub struct ConnectivityCheck<'a, F>
+where
+    F: FnMut(&ObjectId, Kind)
+{
+    /// ODB handle to use for the check
+    db: &'a Handle,
+    /// Closure to invoke when a missing object is encountered
+    missing_cb: F,
+    /// Set of Object IDs already (or about to be) scanned during the check
+    oid_set: HashSet,
+    /// Single buffer for decoding objects from the ODB
+    /// This is slightly faster than allocating throughout the connectivity check (and reduces the memory requirements)
+    buf: Vec<u8>,
+}
+
+impl<'a, F> ConnectivityCheck<'a, F>
+where
+    F: FnMut(&ObjectId, Kind)
+{
+    /// Instantiate a connectivity check
+    pub fn new(db: &'a Handle, missing_cb: F) -> ConnectivityCheck<'a, F> {
+        ConnectivityCheck {
+            db,
+            missing_cb,
+            oid_set: HashSet::default(),
+            buf: Vec::new(),
+        }
+    }
+
+    /// Run the connectivity check on the provided commit object ID
+    /// - This will walk the trees and blobs referenced by the commit and verify they exist in the ODB
+    /// - Any objects previously encountered by this [`ConnectivityCheck`] instance will be skipped
+    /// - Any referenced blobs that are not present in the ODB will result in a call to the  `missing_cb`
+    /// - Missing commits or trees will currently result in panic
+    ///     - TODO: add support for missing trees
+    ///     - TODO: consider how to handle a missing commit (invoke `missing_cb`, or possibly return a Result?)
+    pub fn check_commit(&mut self, oid: &ObjectId) {
```

Comment:
I did try this, but found that because I end up using a `HashSet` of all objects the connectivity check has already encountered, I'll end up needing to create an `ObjectId` out of the `oid` for *every* object anyways. After coming to that conclusion, it seemed to me it'd be more efficient to just convert to an `ObjectId` up-front and use that throughout the connectivity check routines.

I'm probably missing something here... If so, please let me know.

Reply:
Yes, you are right. I also end up using `&ObjectId` when hashsets are involved, I was a bit quick there with the armchair review.